src/Temporalio.Extensions.Hosting/TemporalWorkerService.cs

```diff
@@ -113,27 +114,72 @@ public TemporalWorkerService(
+        /// </summary>
+        /// <param name="serviceId">Unique identifier for the Worker service.</param>
+        /// <param name="optionsMonitor">Used to lookup the options to build the worker with.
+        /// </param>
+        /// <param name="existingClient">Existing client to use if the options don't specify
+        /// client connection options (connected when run if lazy and not connected).</param>
+        /// <param name="loggerFactory">Logger factory to use if not already on the worker options.
+        /// The worker options logger factory or this one will be also be used for the client if an
+        /// existing client does not exist (regardless of client options' logger factory).</param>
+        /// <remarks>
+        /// WARNING: Do not rely on the signature of this constructor, it is for DI container use
+        /// only and may change in incompatible ways.
+        /// </remarks>
+        [ActivatorUtilitiesConstructor]
+        public TemporalWorkerService(
+            TemporalWorkerServiceIdentifier serviceId,
+            IOptionsMonitor<TemporalWorkerServiceOptions> optionsMonitor,
+            ITemporalClient? existingClient = null,
+            ILoggerFactory? loggerFactory = null)
         {
             var options = (TemporalWorkerServiceOptions)optionsMonitor.Get(
                 TemporalWorkerServiceOptions.GetUniqueOptionsName(
-                    taskQueueAndBuildId.TaskQueue, taskQueueAndBuildId.BuildId)).Clone();
+                    serviceId.TaskQueue, serviceId.Version)).Clone();
 
             // Make sure options values match the ones given in constructor
-            if (options.TaskQueue != taskQueueAndBuildId.TaskQueue)
+            if (options.TaskQueue != serviceId.TaskQueue)
             {
                 throw new InvalidOperationException(
-                    $"Task queue '{taskQueueAndBuildId.TaskQueue}' on constructor different than '{options.TaskQueue}' on options");
+                    $"Task queue '{serviceId.TaskQueue}' on constructor different than '{options.TaskQueue}' on options");
             }
-            if (options.BuildId != taskQueueAndBuildId.BuildId)
+
+            if (serviceId.VersionIsBuildId)
             {
-                throw new InvalidOperationException(
-                    $"Build ID '{taskQueueAndBuildId.BuildId ?? "<unset>"}' on constructor different than '{options.BuildId ?? "<unset>"}' on options");
+#pragma warning disable 0618
+                if (options.BuildId != serviceId.Version)
+                {
+                    throw new InvalidOperationException(
+                        $"BuildID '{serviceId.Version ?? "<unset>"}' on constructor different than '{options.BuildId ?? "<unset>"}' on options");
+                }
+#pragma warning restore 0618
+            }
+            else
+            {
+                if (options.DeploymentOptions?.Version?.ToCanonicalString() != serviceId.Version)
```

Comment:
Should there be a check to ensure `options.BuildId` is `null` and `options.UseVersioning` is `false` here? Meaning, if they are using new versioning, shouldn't they not be setting the old values away from their default?

Reply:
I don't think there needs to be, because if the versions match, then they couldn't possibly have set the other options on `options`, because they're exclusive.

Reply:
What if they used the overload that only sets the legacy/deprecated build ID and does not set deployment options, won't this conditional fail for them now?

Reply:
I think my concern was if I was using new deployment options _and_ have manually set old build ID, but I guess worker creation throws for that, no worries